crates/core/tedge_api/goals.md

```diff
@@ -0,0 +1,363 @@
+# Summary
+
+The `tedge_api` crate brings centralized definitions of both messages and
+behaviour by introducing traits for runtime Plugins to hook into. One goal is
+to be *agnostic* on whether a given `Plugin` is part of the same executable or
+not. As all messages are easily de/serializable they are also transport
+agnostic, allowing implementers and users to de/construct them from/for whatever
+sources/destinations they see fit.
+
+# Motivation
+
+Users and developers need to have a common ground to future proof communication
+and growth. A common crate to both will make sure that all sides do not grow
+apart and stay compatible to each other.
+
+# Guide-level explanation
+
+_This section is meant to be read as if the feature was_ already _implemented._
+
+-----
+
+ThinEdge is an edge focused IoT framework. At its core it serves to bridge the
+gap between devices and the cloud. As it is meant to run on low-resources
+devices it's core architecture supports that.
+As such, it is built upon a message passing router with a modular approach for
+extending its functionality through plugins.
+
+ThinEdge can thus be understood as being a common-core of a message passing router
+with a collection of plugins that ultimately define what it actually _does_.
+
+Plugins come in two forms:
+
+- Run-time & external, which are provided to Thin-Edge and meant to
+  interoperate with it specifically through stdin/stdout or some other
+  well-specified interface (cf. external plugins further down)
+    - These are for example custom built executables that a user provides and
+      wishes to not integrate into ThinEdge for various reasons (e.g. language
+      differences or license)
+- Compile-time & built-in, which are compiled into the actual ThinEdge binary
+  and written in Rust.
+    - These offer the advantage of simplifying deployment as well as assurances
+      w.r.t. the messages (i.e. changes)
+
+Both ways are _functionally equivalent_ and _support the same features_.
+
+We aim to provide plugins for all common server and use-cases, see further down
+if you wish to extend Thin-Edge yourself.
+
+## Configuring your plugins
+
+ThinEdge is on its own merely a collection of plugin kinds, and requires a
+configuration to do any useful work. A common configuration includes a cloud
+mapper, some device management plugins as well as some data sources.
+
+> One core idea is that you can create as many plugin instances of a single
+> plugin _kind_ as you wish. For example you could have multiple "File Watcher"
```

Comment:
> What does kind exactly mean here?

I have borrowed the word from [Type Theory](https://en.wikipedia.org/wiki/Kind_(type_theory)) which most succinctly describes it as a `Type Constructor`. Since a `PluginBuilder` can be seen as such, it makes sense to call it a kind.